pglock/core.py

```diff
@@ -0,0 +1,394 @@
+import contextlib
+import datetime as dt
+import functools
+import hashlib
+import inspect
+import threading
+
+from django.apps import apps
+from django.db import connections, DEFAULT_DB_ALIAS, transaction
+from django.db.utils import OperationalError
+import pgactivity
+import psycopg2.extensions
+
+
+# Lock levels
+ACCESS_SHARE = "ACCESS SHARE"
+ROW_SHARE = "ROW SHARE"
+ROW_EXCLUSIVE = "ROW EXCLUSIVE"
+SHARE_UPDATE_EXCLUSIVE = "SHARE UPDATE EXCLUSIVE"
+SHARE = "SHARE"
+SHARE_ROW_EXCLUSIVE = "SHARE ROW EXCLUSIVE"
+EXCLUSIVE = "EXCLUSIVE"
+ACCESS_EXCLUSIVE = "ACCESS EXCLUSIVE"
+
+# Side effects
+RETURN = "RETURN"
+RAISE = "RAISE"
+SKIP = "SKIP"
+
+
+_timeout = threading.local()
+
+
+@contextlib.contextmanager
+def lock_timeout(interval=None, *, using=DEFAULT_DB_ALIAS, **timedelta_kwargs):
+    """Set the lock timeout as a decorator or context manager.
+
+    Nested invocations will successfully apply and rollback the timeout to
+    the previous value.
+
+    Args:
+        timeout (datetime.timedelta, default=None): A timedelta object to specify
+            the timeout interval
+        using (str, default="default"): The database to use.
+        **timedelta_kwargs: Keyword arguments to directly supply to
+            datetime.timedelta to create an interval. E.g.
+            ``pglock.timeout(seconds=1, milliseconds=100)``
+            will create a timeout of 1100 milliseconds.
+
+    Raises:
+        django.db.utils.OperationalError: When a timeout occurs
+    """
+    if interval is None:
+        interval = dt.timedelta(**timedelta_kwargs)
+
+    assert isinstance(interval, dt.timedelta)
+
+    if not hasattr(_timeout, "value"):
```

Comment:
This attribute name might clash with an application's of a local thread attribute.

Reply:
`_timeout` is a private global variable in the module. Not sure what you mean, but this is safe